cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,194 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write;
+
+use indexmap::IndexSet;
+use itertools::Itertools;
+use jj_lib::backend::CommitId;
+use jj_lib::commit::Commit;
+use jj_lib::repo::Repo;
+use jj_lib::revset::{RevsetExpression, RevsetIteratorExt};
+use tracing::instrument;
+
+use crate::cli_util::{short_commit_hash, CommandHelper, RevisionArg};
+use crate::command_error::{user_error, CommandError};
+use crate::commands::rebase::rebase_descendants;
+use crate::ui::Ui;
+
+/// Parallelize revisions by making them siblings
+///
+/// The set of target commits being parallelized must have a single head and
+/// root commit (not to be confused with the repo root), otherwise the command
+/// will fail. Each of the target commits is rebased onto the parents of the
+/// root commit. The children of the head commit are rebased onto the target
+/// commits.
+///
+/// Example usage:
+/// ```text
+/// jj log
+/// ◉   3
+/// ◉   2
+/// ◉   1
+/// ◉
+/// jj parallelize 1::2
+/// jj log
+/// ◉     3
+/// ├─╮
+/// │ ◉   2
+/// ◉ │   1
+/// ├─╯
+/// ◉
+/// ```
+#[derive(clap::Args, Clone, Debug)]
+#[command(verbatim_doc_comment)]
+pub(crate) struct ParallelizeArgs {
+    // #[arg(long, short)]
+    revisions: Vec<RevisionArg>,
+}
+
+#[instrument(skip_all)]
+pub(crate) fn cmd_parallelize(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &ParallelizeArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    // Parse and validate the input revset(s).
+    // Find the parents of each commit. The parent(s) which are ancestors of
+    // each revision are the new parents of the input revisions.
+    // Find the children of each commit. The children which are descendants
+    // of each revision are the new children of the input revisions.
+    // If one or both of these sets cannot be found, then parallelize cannot
+    // proceed.
+    // let target_commits = workspace_command.parse_union_revsets(&args.revisions)?.
+    let mut tx = workspace_command.start_transaction();
+    let target_revset = tx
+        .base_workspace_helper()
+        .parse_union_revsets(&args.revisions)?
+        .expression()
+        .clone();
+    let target_commits: Vec<Commit> = tx
+        .base_workspace_helper()
+        .attach_revset_evaluator(target_revset.clone())?
+        .evaluate()?
+        .iter()
+        // We want parents before children, so we need to reverse the order.
+        .reversed()
+        .commits(tx.base_repo().store())
+        .try_collect::<_, Vec<Commit>, _>()?;
+    if target_commits.len() < 2 {
+        return Ok(writeln!(ui.stderr(), "Nothing to do.")?);
+    }
+    let target_heads: Vec<CommitId> = tx
+        .base_workspace_helper()
+        .attach_revset_evaluator(target_revset.heads())?
```

Comment:
Re to `evaluate_revset()` API stuff:
since `target_revset` expression is evaluated above, a better option is to use `RevsetExpression::commits(commit_ids)` and `evaluate_programmatic()`. That will probably help extract a `&[Commit]` function which doesn't depend on a user revset expression.

Reply:
Ah, yes. Evaluating the user revset once and then constructing a new one from the actual commit IDs solves the issues I had. I want to do a bit more refactoring, but I was able to remove all of the calls to `attach_revset_evaluator` that I had before.